### PR TITLE
add multinode support via slurm trainer, large scale race condition fix

### PR DESCRIPTION
**Pull request by lessw2020**
This PR 
1 - adds multi-node training support via a multinode_trainer.slurm file. Verified llama 7b on 20 nodes / 160 A100s. 
2 - It also corrects a race condition that can occur on larger scale training in profiling, where the check for the trace dir existence fails for process 1, but in the interim another process 2 makes the directory, and then when process 1 tries to make the dir it errors out as the dir exists.
This is a simple fix of adding exist_ok=True to both of the makedir command (dump folder, trace folder). 

Screenshot 2024-02-15 at 10 53 18 PM
Screenshot 2024-02-15 at 10 55 02 PM

### CODE REVIEW

File: multinode_trainer.slurm [hunk 1]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Comment (wconstab) [inline multinode_trainer.slurm:1]: Any info you can add in comments about what cluster env, os/docker, etc. you are assuming might help future us

File: multinode_trainer.slurm [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+#SBATCH --job-name=torchtrain_multi_node
```
Comment (wanchaol) [inline multinode_trainer.slurm:8]: maybe let's add some simple instructions at least in the README.md to document how to use this slurm script in AWS cluster

Comment (lessw2020) [reply multinode_trainer.slurm:8]: added basic info on how to customize for their cluster.

File: multinode_trainer.slurm (hunk 1 above)
Comment (lessw2020) [reply multinode_trainer.slurm:1]: good idea - there's no assumptions about os, but it is optimized towards AWS and EFA, so I've added a few comments in the slurm file as well as usage tips in the readme.  
Longer term might want to expand this to it's own page, as we get feedback from users running larger clusters and any consistent knowledge issues.

**Pull request merged**